apps/rush-lib/src/logic/buildCache/AmazonS3/AmazonS3Client.ts

```diff
@@ -0,0 +1,181 @@
+    const host: string = `${this._s3Bucket}.s3.amazonaws.com`;
+    const signedHeaderNames: string = `${HOST_HEADER_NAME};${CONTENT_HASH_HEADER_NAME};${DATE_HEADER_NAME}`;
+    const canonicalRequest: string = [
+      verb,
+      `/${objectName}`,
+      '', // we don't use query strings for these requests
+      `${HOST_HEADER_NAME}:${host}`,
+      `${CONTENT_HASH_HEADER_NAME}:${bodyHash}`,
+      `${DATE_HEADER_NAME}:${isoDateString.dateTime}`,
+      '',
+      signedHeaderNames,
+      bodyHash
+    ].join('\n');
+    const canonicalRequestHash: string = this._getSha256(canonicalRequest);
+
+    const scope: string = `${isoDateString.date}/${this._s3Region}/s3/aws4_request`;
+    const stringToSign: string = [
+      'AWS4-HMAC-SHA256',
+      isoDateString.dateTime,
+      scope,
+      canonicalRequestHash
+    ].join('\n');
+
+    const dateKey: Buffer = this._getSha256Hmac(`AWS4${this._secretAccessKey}`, isoDateString.date);
+    const dateRegionKey: Buffer = this._getSha256Hmac(dateKey, this._s3Region);
+    const dateRegionServiceKey: Buffer = this._getSha256Hmac(dateRegionKey, 's3');
+    const signingKey: Buffer = this._getSha256Hmac(dateRegionServiceKey, 'aws4_request');
+    const signature: string = this._getSha256Hmac(signingKey, stringToSign, 'hex');
+
+    const authorizationHeader: string = `AWS4-HMAC-SHA256 Credential=${this._accessKeyId}/${scope},SignedHeaders=${signedHeaderNames},Signature=${signature}`;
+
+    const headers: fetch.Headers = new fetch.Headers();
+    headers.set('Authorization', authorizationHeader);
+    headers.set(DATE_HEADER_NAME, isoDateString.dateTime);
+    headers.set(CONTENT_HASH_HEADER_NAME, bodyHash);
+
+    const webFetchOptions: IGetFetchOptions | IPutFetchOptions = {
+      verb,
+      headers
+    };
+    if (verb === 'PUT') {
+      (webFetchOptions as IPutFetchOptions).body = body;
+    }
+
+    const response: fetch.Response = await this._webClient.fetchAsync(
+      `https://${host}/${objectName}`,
+      webFetchOptions
+    );
+
+    return response;
+  }
+
+  public _getSha256Hmac(key: string | Buffer, data: string): Buffer;
+  public _getSha256Hmac(key: string | Buffer, data: string, encoding: 'hex'): string;
+  public _getSha256Hmac(key: string | Buffer, data: string, encoding?: 'hex'): Buffer | string {
+    const hash: crypto.Hash = crypto.createHmac('sha256', key);
+    hash.update(data);
+    if (encoding) {
+      return hash.digest(encoding);
+    } else {
+      return hash.digest();
+    }
+  }
+
+  private _getSha256(data?: string | Buffer): string {
+    if (data) {
+      const hash: crypto.Hash = crypto.createHash('sha256');
+      hash.update(data);
+      return hash.digest('hex');
+    } else {
+      return 'e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855';
+    }
+  }
+
+  private _getIsoDateString(date: Date = new Date()): IIsoDateString {
+    let dateString: string = date.toISOString();
```

Comment:
Show an example value for `dateString` in a code comment, so we know what the `replace()` is supposed to be matching